File: address-geocode-international-dot-net-examples/PlaceSearchRestSdkExample.cs
using address_geocode_international_dot_net.REST;

namespace address_geocode_international_dot_net_examples
{
    internal static class PlaceSearchRestSdkExample
    {
        public static void Go(string licenseKey, bool isLive)
        {
            Console.WriteLine("\r\n------------------------------------------------------");
            Console.WriteLine("Address Geocode International - PlaceSearch - REST SDK");
            Console.WriteLine("------------------------------------------------------");

            // Construct input object
            PlaceSearchClient.PlaceSearchInput placeSearchInput = new(
                SingleLine: "17 Battery Place, New York, NY 10004",
                Address1: "",
                Address2: "",
                Address3: "",
                Address4: "",
                Address5: "",
                Locality: "",
                AdministrativeArea: "",
                PostalCode: "",
                Country: "US",
                Boundaries: "",
                MaxResults: "2",
                SearchType: "Address",
                Extras: "",
                LicenseKey: licenseKey,
                IsLive: isLive,
                TimeoutSeconds: 15
            );

            Console.WriteLine("\r\n* Input *\r\n");
            Console.WriteLine($"Single Line       : {placeSearchInput.SingleLine}");
            Console.WriteLine($"Address1          : {placeSearchInput.Address1}");
            Console.WriteLine($"Address2          : {placeSearchInput.Address2}");
            Console.WriteLine($"Address3          : {placeSearchInput.Address3}");
            Console.WriteLine($"Address4          : {placeSearchInput.Address4}");
            Console.WriteLine($"Address5          : {placeSearchInput.Address5}");
            Console.WriteLine($"Locality          : {placeSearchInput.Locality}");
            Console.WriteLine($"AdministrativeArea: {placeSearchInput.AdministrativeArea}");
            Console.WriteLine($"PostalCode        : {placeSearchInput.PostalCode}");
            Console.WriteLine($"Country           : {placeSearchInput.Country}");
            Console.WriteLine($"Boundaries        : {placeSearchInput.Boundaries}");
            Console.WriteLine($"MaxResults        : {placeSearchInput.MaxResults}");
            Console.WriteLine($"SearchType        : {placeSearchInput.SearchType}");
            Console.WriteLine($"Extras            : {placeSearchInput.Extras}");
            Console.WriteLine($"Is Live           : {placeSearchInput.IsLive}");
            Console.WriteLine($"License Key       : {placeSearchInput.LicenseKey}");
            Console.WriteLine($"Timeout (sec)     : {placeSearchInput.TimeoutSeconds}");

            // Invoke REST API
            AGIPlaceSearchResponse response = PlaceSearchClient.Invoke(placeSearchInput);

            if (response.Error is null)
            {
                Console.WriteLine("\r\n* Results *\r\n");

                if (response.SearchInfo != null)
                {
                    Console.WriteLine("Search Info:");
                    Console.WriteLine($"Status           : {response.SearchInfo.Status}");
                    Console.WriteLine($"NumberOfLocations: {response.SearchInfo.NumberOfLocations}");
                    Console.WriteLine($"Notes            : {response.SearchInfo.Notes}");
                    Console.WriteLine($"NotesDesc        : {response.SearchInfo.NotesDesc}");
                    Console.WriteLine($"Warnings         : {response.SearchInfo.Warnings}");
                    Console.WriteLine($"WarningDesc      : {response.SearchInfo.WarningDesc}");

                    Console.WriteLine();
                }

                if (response.Locations != null && response.Locations.Length > 0)
                {
                    int count = 1;
                    foreach (var location in response.Locations)
                    {
                        Console.WriteLine($"\r\nLocation #{count++}\r\n");
                        Console.WriteLine($"\tPrecision Level: {location.PrecisionLevel}");
                        Console.WriteLine($"\tType           : {location.Type}");
                        Console.WriteLine($"\tLatitude       : {location.Latitude}");
                        Console.WriteLine($"\tLongitude      : {location.Longitude}");

                        if (location.AddressComponents != null)
                        {
                            Console.WriteLine($"\tPremiseNumber                  : {location.AddressComponents.PremiseNumber}");
                            Console.WriteLine($"\tThoroughfare                   : {location.AddressComponents.Thoroughfare}");
                            Console.WriteLine($"\tDoubleDependentLocality        : {location.AddressComponents.DoubleDependentLocality}");
                            Console.WriteLine($"\tDependentLocality              : {location.AddressComponents.DependentLocality}");
                            Console.WriteLine($"\tLocality                       : {location.AddressComponents.Locality}");
                            Console.WriteLine($"\tAdministrativeArea1            : {location.AddressComponents.AdministrativeArea1}");
                            Console.WriteLine($"\tAdministrativeArea1Abbreviation: {location.AddressComponents.AdministrativeArea1Abbreviation}");
                            Console.WriteLine($"\tAdministrativeArea2            : {location.AddressComponents.AdministrativeArea2}");
                            Console.WriteLine($"\tAdministrativeArea2Abbreviation: {location.AddressComponents.AdministrativeArea2Abbreviation}");
                            Console.WriteLine($"\tAdministrativeArea3            : {location.AddressComponents.AdministrativeArea3}");
                            Console.WriteLine($"\tAdministrativeArea3Abbreviation: {location.AddressComponents.AdministrativeArea3Abbreviation}");
                            Console.WriteLine($"\tAdministrativeArea4            : {location.AddressComponents.AdministrativeArea4}");
                            Console.WriteLine($"\tAdministrativeArea4Abbreviation: {location.AddressComponents.AdministrativeArea4Abbreviation}");
                            Console.WriteLine($"\tPostalCode                     : {location.AddressComponents.PostalCode}");
                            Console.WriteLine($"\tCountry                        : {location.AddressComponents.Country}");
                            Console.WriteLine($"\tCountryISO2                    : {location.AddressComponents.CountryISO2}");
                            Console.WriteLine($"\tCountryISO3                    : {location.AddressComponents.CountryISO3}");
                            Console.WriteLine($"\tPlaceName                      : {location.AddressComponents.PlaceName}");
                            Console.WriteLine($"\tGoogleMapsURL                  : {location.AddressComponents.GoogleMapsURL}");
                            Console.WriteLine($"\tIsUnincorporated               : {location.AddressComponents.IsUnincorporated}");
                            Console.WriteLine($"\tStateFIPS                      : {location.AddressComponents.StateFIPS}");
                            Console.WriteLine($"\tCountyFIPS                     : {location.AddressComponents.CountyFIPS}");
                            Console.WriteLine($"\tCensusTract                    : {location.AddressComponents.CensusTract}");
                            Console.WriteLine($"\tCensusBlock                    : {location.AddressComponents.CensusBlock}");
                            Console.WriteLine($"\tCensusGeoID                    : {location.AddressComponents.CensusGeoID}");
                            Console.WriteLine($"\tClassFP                        : {location.AddressComponents.ClassFP}");
                            Console.WriteLine($"\tCongressCode                   : {location.AddressComponents.CongressCode}");
                            Console.WriteLine($"\tSLDUST                         : {location.AddressComponents.SLDUST}");
                            Console.WriteLine($"\tSLDLST                         : {location.AddressComponents.SLDLST}");
                            Console.WriteLine($"\tTimezone_UTC                   : {location.AddressComponents.TimeZone_UTC}");
                        }

                    }
                }
                else
                {
                    Console.WriteLine("No locations found.");
                }
            }
            else
            {
                Console.WriteLine("\r\n* Error *\r\n");

                Console.WriteLine($"Error Type     : {response.Error.Type}");
                Console.WriteLine($"Error Type Code: {response.Error.TypeCode}");
                Console.WriteLine($"Error Desc     : {response.Error.Desc}");
                Console.WriteLine($"Error Desc Code: {response.Error.DescCode}");
            }
        }
    }
}

File: address-geocode-international-python-examples/place_search_rest_sdk_example.py
import sys
import os

# Add REST client path for module import
sys.path.insert(0, os.path.abspath("../address-geocode-international-python/REST"))

from place_search_rest import place_search

def place_search_rest_sdk_go(license_key, is_live):
    print("\n" + "-" * 54)
    print("Address Geocode International - PlaceSearch - REST SDK")
    print("-" * 54)

    print("* Input *\n")
    print(f"SingleLine         : 17 Battery Place, New York, NY 10004")
    print(f"Address1           : ")
    print(f"Address2           : ")
    print(f"Address3           : ")
    print(f"Address4           : ")
    print(f"Address5           : ")
    print(f"Locality           : ")
    print(f"Administrative Area: ")
    print(f"Postal Code        : ")
    print(f"Country            : USA")
    print(f"Boundaries         : ")
    print(f"Max Results        : ")
    print(f"Search Type        : ")
    print(f"Extras             : ")
    print(f"Is Live            : {is_live}")
    print(f"License Key        : {license_key}")

    try:
        response = place_search(
            single_line="17 Battery Place, New York, NY 10004",
            address1="",
            address2="",
            address3="",
            address4="",
            address5="",
            locality="",
            administrative_area="",
            postal_code="",
            country="USA",
            boundaries="",
            max_results="",
            search_type="",
            extras="",
            license_key=license_key,
            is_live=is_live
        )

        if not response.get('Error'):
            print("\n* Results *\n")

            if response.get('SearchInfo'):
                info = response.get("SearchInfo", {})
            
                print("Search Info:")
                print(f"Status           : {info.get('Status', '')}")
                print(f"NumberOfLocations: {info.get('NumberOfLocations', '')}")
                print(f"Notes            : {info.get('Notes', '')}")
                print(f"NotesDesc        : {info.get('NotesDesc', '')}")
                print(f"Warnings         : {info.get('Warnings', '')}")
                print(f"WarningDesc      : {info.get('WarningDesc', '')}\n")

                if response.get('Locations'):
                    locations = response.get("Locations", [])
                    for idx, loc in enumerate(locations, start=1):
                        print(f"Location #{idx}")
                        print(f"\tPrecisionLevel: {loc.get('PrecisionLevel', '')}")
                        print(f"\tType          : {loc.get('Type', '')}")
                        print(f"\tLatitude      : {loc.get('Latitude', '')}")
                        print(f"\tLongitude     : {loc.get('Longitude', '')}")

                        if loc.get('AddressComponents'):
                            address = loc.get("AddressComponents", {})
                            if address:
                                print(f"\tPremiseNumber                  : {address.get('PremiseNumber', '')}")
                                print(f"\tThoroughfare                   : {address.get('Thoroughfare', '')}")
                                print(f"\tDoubleDependentLocality        : {address.get('DoubleDependentLocality', '')}")
                                print(f"\tDependentLocality              : {address.get('DependentLocality', '')}")
                                print(f"\tLocality                       : {address.get('Locality', '')}")
                                print(f"\tAdministrativeArea1            : {address.get('AdministrativeArea1', '')}")
                                print(f"\tAdministrativeArea1Abbreviation: {address.get('AdministrativeArea1Abbreviation', '')}")
                                print(f"\tAdministrativeArea2            : {address.get('AdministrativeArea2', '')}")
                                print(f"\tAdministrativeArea2Abbreviation: {address.get('AdministrativeArea2Abbreviation', '')}")
                                print(f"\tAdministrativeArea3            : {address.get('AdministrativeArea3', '')}")
                                print(f"\tAdministrativeArea3Abbreviation: {address.get('AdministrativeArea3Abbreviation', '')}")
                                print(f"\tAdministrativeArea4            : {address.get('AdministrativeArea4', '')}")
                                print(f"\tAdministrativeArea4Abbreviation: {address.get('AdministrativeArea4Abbreviation', '')}")
                                print(f"\tPostalCode                     : {address.get('PostalCode', '')}")
                                print(f"\tCountry                        : {address.get('Country', '')}")
                                print(f"\tCountryISO2                    : {address.get('CountryISO2', '')}")
                                print(f"\tCountryISO3                    : {address.get('CountryISO3', '')}")
                                print(f"\tPlaceName                      : {address.get('PlaceName', '')}")
                                print(f"\tGoogleMapsURL                  : {address.get('GoogleMapsURL', '')}")
                                print(f"\tIsUnincorporated               : {address.get('IsUnincorporated', '')}")
                                print(f"\tStateFIPS                      : {address.get('StateFIPS', '')}")
                                print(f"\tCountyFIPS                     : {address.get('CountyFIPS', '')}")
                                print(f"\tCensusTract                    : {address.get('CensusTract', '')}")
                                print(f"\tCensusBlock                    : {address.get('CensusBlock', '')}")
                                print(f"\tCensusGeoID                    : {address.get('CensusGeoID', '')}")
                                print(f"\tClassFP                        : {address.get('ClassFP', '')}")
                                print(f"\tCongressCode                   : {address.get('CongressCode', '')}")
                                print(f"\tSLDUST                         : {address.get('SLDUST', '')}")
                                print(f"\tSLDLST                         : {address.get('SLDLST', '')}")
                                print(f"\tTimeZone_UTC                   : {address.get('TimeZone_UTC', '')}")
        else:
            print("\n* Error *\n")
            err = response.get('Error', {})
            print(f"\tError Type     : {err.get('Type', '')}")
            print(f"\tError Type Code: {err.get('TypeCode', '')}")
            print(f"\tError Desc     : {err.get('Desc', '')}")
            print(f"\tError Desc Code: {err.get('DescCode', '')}")
    except RuntimeError as ex:
        print("\nError calling service:", ex)
